### PR TITLE
Add ArrayBuffer with user specified buffer

### PR DESCRIPTION
**Pull request by galpeter**
New API functions:
  - jerry_create_arraybuffer_external
  - jerry_arraybuffer_get_pointer

### CODE REVIEW

**Comment by galpeter** [comment]
Depends on: #2161

**Comment by LaszloLango** [comment]
#2161 landed. Please rebase.

**Comment by galpeter** [comment]
PR rebased

File: jerry-core/api/jerry.c [hunk 1]
```diff
@@ -2798,6 +2834,45 @@ jerry_get_arraybuffer_byte_length (const jerry_value_t value) /**< ArrayBuffer *
...
 } /* jerry_get_arraybuffer_byte_length */
 
+/**
+ * Get a pointer for the start of the ArrayBuffer.
+ *
+ * Note:
+ *    * Only valid for ArrayBuffers created with jerry_create_arraybuffer_external.
+ *    * This is a high-risk operation as the bounds are not checked
+ *      when accessing the pointer elements.
+ *    * jerry_release_value must be called when the pointer is no longer needed.
```
Comment (zherczeg) [inline jerry-core/api/jerry.c]: release what?

File: jerry-core/ecma/base/ecma-globals.h [hunk 2]
```diff
@@ -1243,6 +1245,32 @@ typedef struct
 
 #ifndef CONFIG_DISABLE_ES2015_TYPEDARRAY_BUILTIN
 
+/**
+ * Extra information for ArrayBuffers.
+ */
+typedef enum
+{
+  ECMA_ARRAYBUFFER_INTERNAL_MEMORY = 0u,        /* ArrayBuffer memory is handled internally. */
+  ECMA_ARRAYBUFFER_EXTERNAL_MEMORY = (1u << 0), /* Memory provided by the developer. */
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-globals.h]: Comment: ArrayBuffer created by jerry_create_arraybuffer_external.

File: jerry-core/ecma/base/ecma-globals.h [hunk 3]
```diff
@@ -1243,6 +1245,32 @@ typedef struct
...
+} ecma_arraybuffer_extra_flag_t;
+
+#define ECMA_ARRAYBUFFER_HAS_EXTERNAL_MEMORY(object_p) \
+    ((((ecma_extended_object_t *) object_p)->u.class_prop.extra_info & ECMA_ARRAYBUFFER_EXTERNAL_MEMORY) == 1)
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-globals.h]: `!= 0` is nicer than `== 1`

File: jerry-core/ecma/base/ecma-globals.h [hunk 4]
```diff
@@ -1243,6 +1245,32 @@ typedef struct
...
+ * Struct to store information for ArrayBuffers with external memory.
+ *
+ * The following elements are stored in Jerry memory.
+ *
+ *  buffer_p - pointer to the external memory.
+ *  free_cb - pointer to a callback function which is called when the ArrayBuffer is freed.
+ */
+typedef struct
+{
+  void *buffer_p; /**< external buffer pointer */
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-globals.h]: It is a better technique if the "base" structure is used as the first member. Reason: this ensures proper alignment for structure members.

**Comment by galpeter** [comment]
@zherczeg I've updated the PR based on the requests.

File: docs/02.API-REFERENCE.md [hunk 5]
```diff
@@ -4921,3 +4972,61 @@ jerry_arraybuffer_write (const jerry_value_t value,
 - [jerry_create_arraybuffer](#jerry_create_arraybuffer)
 - [jerry_arraybuffer_write](#jerry_arraybuffer_write)
 - [jerry_get_arraybuffer_byte_length](#jerry_get_arraybuffer_byte_length)
+
+
+## jerry_get_arraybuffer_pointer
+
+**Summary**
+
+The function allows acces to the contents of the Array Buffer directly.
```
Comment (LaszloLango) [inline docs/02.API-REFERENCE.md]: `access`

File: docs/02.API-REFERENCE.md [hunk 6]
```diff
@@ -4921,3 +4972,61 @@ jerry_arraybuffer_write (const jerry_value_t value,
...
+Only allowed for Array Buffers which were created with
+[jerry_create_arraybuffer_external](#jerry_create_arraybuffer_external)
+function calls. In any other case this function will return `NULL`.
+
+After using the pointer the [jerry_release_value](#jerry_release_value)
+function must be called.
+
+WARNING! This operation is for expert use only! The programmer must
```
Comment (LaszloLango) [inline docs/02.API-REFERENCE.md]: Please make the warning text bold: `**WARNING!**`

File: docs/02.API-REFERENCE.md [hunk 7]
```diff
@@ -2465,6 +2466,56 @@ jerry_create_arraybuffer (jerry_length_t size);
 - [jerry_release_value](#jerry_release_value)
 
 
+## jerry_create_arraybuffer_external
+
+**Summary**
+
+Create a jerry_value_t representing an ArrayBuffer object with
```
Comment (LaszloLango) [inline docs/02.API-REFERENCE.md]: `Creates`

File: jerry-core/ecma/operations/ecma-arraybuffer-object.c [hunk 8]
```diff
@@ -60,6 +61,38 @@ ecma_arraybuffer_new_object (ecma_length_t length) /**< length of the arraybuffe
...
 
+/**
+ * Helper function: create arraybuffer object with external buffer backing.
+ *
+ * The struct of external arraybuffer object:
+ *   ecma_object_t
+ *   extend_part
+ *   arraybuffer external info part
+ *
+ * @return ecma_object_t *
```
Comment (LaszloLango) [inline jerry-core/ecma/operations/ecma-arraybuffer-object.c]: `@return Pointer to the created ArrayBuffer object if success, NULL otherwise.`

Comment (galpeter) [reply jerry-core/ecma/operations/ecma-arraybuffer-object.c]: NULL can't be returned here

**Comment by galpeter** [comment]
@LaszloLango I've updated the PR based on the requests.

**Pull request merged**